### PR TITLE
FIX: subtle race condition in cbook.mkdirs

### PR DESCRIPTION
**Pull request by tacaswell**
attn @timj

### CODE REVIEW

**Comment by tacaswell** [comment]
closes #5225

**Comment by timj** [comment]
Thanks. Looks good to me except that I wonder why the Python2.7 code calls `makedirs` when it's not really using any of the functionality of that routine and may as well be using `mkdir` for each path component.

**Comment by tacaswell** [comment]
Good point, simplified.

File: lib/matplotlib/cbook.py [hunk 1]
```diff
@@ -1012,18 +1012,26 @@ def mkdirs(newdir, mode=0o777):
...
-                    os.makedirs(thispart, mode)
-
-    except OSError as err:
-        # Reraise the error unless it's about an already existing directory
-        if err.errno != errno.EEXIST or not os.path.isdir(newdir):
-            raise
+    # this functionality is now in core python as of 3.2
+    if six.PY3:
+        os.makedirs(newdir, mode=mode, exist_ok=True)
+        return
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: (nit): I know it's equivalent, but I think I'd prefer to see the Python 2 stuff in an `else` clause so it's more obvious it only runs on Python 2.

Comment (tacaswell) [inline lib/matplotlib/cbook.py]: done.  That make lots of sense and will make ripping out the LPY stuff easier.

File: lib/matplotlib/cbook.py [hunk 2]
```diff
@@ -1012,20 +1012,18 @@ def mkdirs(newdir, mode=0o777):
...
-            parts = os.path.split(newdir)
-            for i in range(1, len(parts) + 1):
-                thispart = os.path.join(*parts[:i])
-                if not os.path.exists(thispart):
-                    os.makedirs(thispart, mode)
-
-    except OSError as err:
-        # Reraise the error unless it's about an already existing directory
-        if err.errno != errno.EEXIST or not os.path.isdir(newdir):
+        os.makedirs(newdir, mode=mode)
```
Comment (timj) [inline lib/matplotlib/cbook.py]: Do we know that this works with multiple processes? I thought the whole point of `exist_ok` in Python3 was to handle the case of the directory pre-existing, and I thought the initial attempt at rewriting the code to look at each leaf separately was to work around the multi-process problem. Isn't this code now back to how it was before the change to protect from multiple processes attempting to create the directory? My previous comment was that `os.makedirs` looked like it should just be `os.mkdir` as the code was only ever creating a directory at a time.

**Comment by timj** [comment]
The reimplementation of `makedirs` seemed to appear in 529c88e7369b519b2c7e49bf4034f4da964f08cb but the commit message doesn't really tell me anything about why that change was made.

File: lib/matplotlib/cbook.py (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/cbook.py]: I think we have now delegated that problem down to cpython?

**Comment by timj** [comment]
@tacaswell for some reason I can't see the comment I just got in email on Github. I think I'm saying that CPython 2.7 may not handle the multi-process race condition properly and simply calling `os.makedirs` may not be the real fix. I'm not sure if that's the case though as I assume the code in 529c88e7369b519b2c7e49bf4034f4da964f08cb was added for a reason. Maybe Python 2.7 doesn't have the problem (the code was put in to matplotlib in 2008)?

**Comment by tacaswell** [comment]
https://github.com/python/cpython/blob/master/Lib/os.py#L216 <- current implementation of `os.makedirs` which looks like it handles things properly.

https://github.com/python/cpython/blob/2.7/Lib/os.py#L136 <- implementation for 2.7 which also looks like it handles this race condition properly.

It looks like 2.5 (https://github.com/python/cpython/blob/2.5/Lib/os.py#L150) is where that checking first shows up (2.4: https://github.com/python/cpython/blob/2.4/Lib/os.py#L143)


you have to click the 'show outdated diff' link up the page a bit to see the comment.

**Comment by timj** [comment]
Thanks. I'm happy now :smile: :+1:

**Pull request merged**